### PR TITLE
Facebook Pages Support

### PR DESCRIPTION
**Pull request by hokiepokedad**
# Description

### Required Parameters
These `alarms.json` parameters - `active`, `type`, and `page_access_token` - are required to enable the Facebook Pages alarm service:

| Parameters         | Description                            |
| --------------     |----------------------------------------|
| type               | must be `facebookpages`                |
| active             | `True` for alarm to be active          |
| page_access_token  | Your permanent FB page access token    |

### Example: Basic Alarm Configuration using Required Parameters
```json
{
	"active": "True",
	"type":"facebookpages",
	"page_access_token":"YOUR_PAGE_ACCESS_TOKEN"
}
```
**Note:** The above code is to be inserted into the alarms section of alarms.json. It does not represent the entire alarms.json file.

### Optional Parameters
In addition to the 3 required parameters, several optional parameters are available to personalize your Facebook Pages notifications.  Below is an example of these optional parameters and how they are incorporated into a functional alarm layout for Facebook Pages.

These optional parameters, `startup_message`, and `startup_list`, are entered at the same level as `"type":"facebookpages"`.

| Parameters         | Description                                                | Default                      |
|--------------------|------------------------------------------------------------|------------------------------|
| `startup_message`  | confirmation post when PokeAlarm initialized               | `True`                       |
| `startup_list`     | First post will list all alarmed pokemon enabled in `alarms.json`    | `True`            |

These optional parameters below are applicable to the `pokemon`, `pokestop`, and `gym` sections of the JSON file.

| Parameters       | Description                                       | Default                                       |
| -----------------|---------------------------------------------------|-----------------------------------------------|
| `message`        | Message to post as status message                 | `A wild <pkmn> has appeared! Available until <24h_time> (<time_left>).`                                      | 
| `link`           | Link to be added to notification text   		   | `<gmaps>`                                     |

*Note: Nidorans will be `nidoranf` or `nidoranm`, Farfetch'd will be `farfetchd`, and Mr. Mime will be `mrmime`.

## Example: Alarm Configuration Using Optional Parameters
```json
{
	"active": "True",
	"type":"facebookpages",
	"page_access_token":"YOUR_PAGE_ACCESS_TOKEN",
	"startup_message":"True",
	"startup_list":"True",
	"pokemon":{
		"message": "<pkmn> available. <move1>/<move2> (<iv>% - <atk>/<def>/<sta>)",
		"link": "<gmaps>"
	},
	"pokestop":{
		"message": "Someone has placed a lure on a Pokestop! Lure will expire at <24h_time> (<time_left>).",
		"link": "<gmaps>"
	},
	"gym":{
		"message":"A Team <old_team> gym has fallen! It is now controlled by <new_team>.",
		"link": "<gmaps>"
	}
}
```
**Note:** The above code is to be inserted into the alarms section of alarms.json. It does not represent the entire alarms.json file.

 
## How to get a Permanent Facebook Page Token

Use the [Graph API Explorer](https://developers.facebook.com/tools/explorer) for all of these steps except where otherwise stated.

### 0. Create Facebook App

**If you already have an app**, skip to step 1.

1. Go to [My Apps](https://developers.facebook.com/apps/).
2. Click "+ Add a New App".
3. Setup a website app.
4. Go to App Review of your new app and make sure your app is made Public.

### 1. Get User Short-Lived Access Token

1. Go to the [Graph API Explorer](https://developers.facebook.com/tools/explorer).
2. Select the application you want to get the access token for by clicking on the "Graph API Explorer" dropdown near the top right.
3. Click "Get Token" dropdown and choose "Get User Access Token".
4. In the pop-up, under the "Events, Groups & Pages" section, check "manage_pages" and "publish_pages".
5. Click "Get Access Token".
6. Grant access from a Facebook account that has access to manage the target page. Note that if this user loses access the final, never-expiring access token will likely stop working.

Token that appears in the "Access Token" field is your short-lived access token.

### 2. Generate Long-Lived Access Token

Following [these instructions](https://developers.facebook.com/docs/facebook-login/access-tokens#extending) from the Facebook docs, make a GET request to

`https://graph.facebook.com/v2.2/oauth/access_token?grant_type=fb_exchange_token&client_id={app_id}&client_secret={app_secret}&fb_exchange_token={short_lived_token}`

entering in your app's ID and secret, generated from step 0, and the short-lived token, generated in the previous step. Be sure to remove the curly braces "{}" when replacing values.

You **cannot use the Graph API Explorer**. For some reason it gets stuck on this request. I think it's because the response isn't JSON, but a query string. Since it's a GET request, you can just go to the URL in your browser.

The response should look like this:

`access_token=ABC123&expires=5182959`

"ABC123" will be your long-lived access token. You can put it into the [Access Token Debugger](https://developers.facebook.com/tools/debug/accesstoken) to verify. Under "Expires" it should have something like "2 months".

### 4. Get User ID

Using the long-lived access token, make a GET request to

`https://graph.facebook.com/v2.2/me?access_token={long_lived_access_token}`

Be sure to remove the curly braces "{}" when replacing values. The **id** field is your account ID. You'll need it for the next step.

### 5. Get Permanent Page Access Token

Make a GET request to

`https://graph.facebook.com/v2.2/{account_id}/accounts?access_token={long_lived_access_token}`

Be sure to remove the curly braces "{}" when replacing values. The JSON response should have a **data** field under which is an array of items the user has access to. Find the item for the page you want the permanent access token from. The **access_token** field should have your permanent access token. Copy it and test it in the [Access Token Debugger](https://developers.facebook.com/tools/debug/accesstoken). Under "Expires" it should say "Never".

# How Has This Been Tested?

Tested this myself with 14 FB pages.

# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
# Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: alarms/FacebookPages/facebookpages_alarm.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	#Establish connection with FacebookPages
+	def connect(self):
+		self.client = facebook.GraphAPI(self.page_access_token)
+			
+	#Set the appropriate settings for each alert
+	def set_alert(self, settings, default):
+		alert = {}
+		alert['message'] = settings.get('message', default['message'])
+		
+		if parse_boolean(settings.get('location', default['location'])):
```
Comment (kvangent) [inline alarms/FacebookPages/facebookpages_alarm.py]: Perhaps you should add the field 'location' to a customizable field called link instead so users can change the link if needed (if they want to customize it to their map or anything like that)

File: alarms/alarm_manager.py [hunk 2]
```diff
@@ -213,12 +213,11 @@ def trigger_pokemon(self, pkmn):
 
 		pkmn_info = {
 			'id': str(pkmn_id),
-			'encounter_id': pkmn['encounter_id'],
  			'pkmn': name,
 			'lat' : "{}".format(repr(lat)),
 			'lng' : "{}".format(repr(lng)),
 			'gmaps': get_gmaps_link(lat, lng),
-			'dist': get_dist_str(dist),
+			'dist': "%d%s" % (dist, 'yd' if config['UNITS'] == 'imperial' else 'm'),
```
Comment (kvangent) [inline alarms/alarm_manager.py]: Was there a reason you ditched get_dist_str?

File: alarms/FacebookPages/facebookpages_alarm.py (hunk 1 above)
Comment (hokiepokedad) [inline alarms/FacebookPages/facebookpages_alarm.py]: @kvangent yes good idea.  doing that now.

File: alarms/alarm_manager.py (hunk 2 above)
Comment (hokiepokedad) [inline alarms/alarm_manager.py]: nice catch, no reason.  i possibly had an older version of alarm_manager.py.  reverting now.

**Comment by hokiepokedad** [comment]
@kvangent changes made per your comments

**Comment by hokiepokedad** [comment]
@kvangent PR submitted for the Facebook Pages wiki.

**Comment by Imoq** [comment]
This is working beautifully. @hokiepokedad spent a good amount of time helping me on Facebook while adjusting documentation that needed some clarification. It works perfectly on my FB page. **Thank you very much!**

File: alarms/FacebookPages/facebookpages_alarm.py [hunk 3]
```diff
@@ -56,20 +56,16 @@ def connect(self):
 	def set_alert(self, settings, default):
 		alert = {}
 		alert['message'] = settings.get('message', default['message'])
-		
-		if parse_boolean(settings.get('location', default['location'])):
-			alert['attachment'] = { 'link': "<gmaps>" }
+		alert['attachment'] = { 'link': settings.get('link', default['link']) }
```
Comment (kvangent) [inline alarms/FacebookPages/facebookpages_alarm.py]: Is there a reason why you left this as a dict? I think that just making it `alert['link'] = settings.get('link', default['link']) would be more inline with the rest of Alarms.

Comment (hokiepokedad) [inline alarms/FacebookPages/facebookpages_alarm.py]: This was already fixed in a later commit.

**Comment by kvangent** [comment]
@hokiepokedad I don't know if I just missed it the first time around but why did you remove line 216 in Alarm Manager?

**Comment by hokiepokedad** [comment]
@kvangent this was also unintentional and due to making changes to an older version of PokeAlarm.  I'll fix it in a few.

**Comment by hokiepokedad** [comment]
@kvangent this has been corrected now.

**Comment by fufufabulous** [comment]
It's working since few days perfectly

**Pull request merged**